shell/platform/windows/keyboard_manager_win32.cc

```diff
@@ -338,38 +252,64 @@ bool KeyboardManagerWin32::HandleMessage(UINT const action,
       if (wparam == VK_PACKET) {
         return false;
       }
+
+      const uint8_t scancode = (lparam >> 16) & 0xff;
+      const bool extended = ((lparam >> 24) & 0x01) == 0x01;
+      // If the key is a modifier, get its side.
+      const uint16_t key_code = ResolveKeyCode(wparam, extended, scancode);
+      const bool was_down = lparam & 0x40000000;
+
+      // Detect a pattern of key events in order to forge a CtrlLeft up event.
+      // See |IsKeyDownAltRight| for explanation.
+      if (IsKeyDownAltRight(action, key_code, extended)) {
+        if (last_key_is_ctrl_left_down) {
+          should_synthesize_ctrl_left_up = true;
+        }
+      }
+      if (IsKeyDownCtrlLeft(action, key_code)) {
+        last_key_is_ctrl_left_down = true;
+        ctrl_left_scancode = scancode;
+        should_synthesize_ctrl_left_up = false;
+      } else {
+        last_key_is_ctrl_left_down = false;
+      }
+      if (IsKeyUpAltRight(action, key_code, extended)) {
+        if (should_synthesize_ctrl_left_up) {
+          should_synthesize_ctrl_left_up = false;
+          const LPARAM lParam =
+              (1 /* repeat_count */ << 0) | (ctrl_left_scancode << 16) |
+              (0 /* extended */ << 24) | (1 /* prev_state */ << 30) |
+              (1 /* transition */ << 31);
+          window_delegate_->Win32DispatchMessage(WM_KEYUP, VK_CONTROL, lParam);
+        }
+      }
+
       current_session_.clear();
       current_session_.push_back(
           Win32Message{.action = action, .wparam = wparam, .lparam = lparam});
       const bool is_keydown_message =
           (action == WM_KEYDOWN || action == WM_SYSKEYDOWN);
-      // Check if this key produces a character. If so, the key press should
-      // be sent with the character produced at WM_CHAR. Store the produced
-      // keycode (it's not accessible from WM_CHAR) to be used in WM_CHAR.
-      //
-      // Messages with Control or Win modifiers down are never considered as
-      // character messages. This allows key combinations such as "CTRL + Digit"
-      // to properly produce key down events even though `MapVirtualKey` returns
-      // a valid character. See https://github.com/flutter/flutter/issues/85587.
+      // Check if this key produces a character by peeking if this key down
+      // message has a following char message. Certain key messages are not
+      // followed by char messages even though `MapVirtualKey` returns a valid
+      // character (such as Ctrl + Digit, see
+      // https://github.com/flutter/flutter/issues/85587 .)
```

Comment:
Super-ultra nit: :-)

```suggestion
      // https://github.com/flutter/flutter/issues/85587 ).
```